src/meta/src/hummock/compaction_scheduler.rs

```diff
@@ -137,127 +148,344 @@ where
         tracing::info!("Compaction scheduler is stopped");
     }
 
+    /// Tries to pick a compaction task, schedule it to a compactor.
+    ///
+    /// Returns true if a task is successfully picked and sent.
     async fn pick_and_assign(
         &self,
         compaction_group: CompactionGroupId,
         request_channel: Arc<CompactionRequestChannel>,
-    ) -> bool {
+    ) -> ScheduleStatus {
+        let schedule_status = self
+            .pick_and_assign_impl(compaction_group, request_channel)
+            .await;
+        let cancel_task = match &schedule_status {
+            ScheduleStatus::Ok => None,
+            ScheduleStatus::NoTask => None,
+            ScheduleStatus::NoAvailableCompactor(task) => Some(task.clone()),
+            ScheduleStatus::AssignFailure(task) => Some(task.clone()),
+            ScheduleStatus::SendFailure(task) => Some(task.clone()),
+            ScheduleStatus::PickFailure => None,
+        };
+        if let Some(mut compact_task) = cancel_task {
+            // Try to cancel task immediately.
+            if self
+                .hummock_manager
+                .cancel_compact_task(&mut compact_task)
+                .await
+                .is_err()
+            {
```

Comment:
we cancel task asynchronously, how can we track it if `cancel_compact_task` fail? Maybe need some log / metrics here ?

Reply:
Fixed.

Reply:
We log whenever cancel fails, either here or in the async worker.